### PR TITLE
Add p11tool and p11-kit support

### PR DESCRIPTION
**Pull request by davidcassany**
This commit adds support to retrieve trusted certificates using
p11tool and p11-kit. This provides a better cross distro support,
however it requires the distro to support it, because of that
the path guest strategy is kept as a fallback. So the behavior
remains untouched for hosts not supporting p11-kit.

Another major benefit of p11-kit approach is that it enables to
read host certificates from a socket, this is specially relevant
in container contexts where one might want to use host certificates
inside a containerized application.

Signed-off-by: David Cassany <dcassany@suse.com>

### CODE REVIEW

**Comment by davidcassany** [comment]
@kanishk98 gentle ping, is there any chance to move on with this PR?

**Comment by davidcassany** [comment]
@kanishk98 is this repository still alive? Can I help somehow to move/discuss this PR forward?

**Comment by kanishk98** [comment]
@davidcassany I'm incredibly sorry for the delay in getting back to you, it's been a while since I checked out what was going on with this repo. If you don't mind, I'll take till later tonight IST to get back on your PRs (#2 looks good for sure, will just test it out locally once) and I'm afraid I'll have to educate myself a little bit on #3. I hope that's alright. 
Thanks a ton for helping out here, and once again, really sorry for the very long delay.

**Comment by davidcassany** [comment]
> I'm incredibly sorry for the delay in getting back to you, it's been a while since I checked out what was going on with this repo

No problem, thanks for looking at it.

> I'm afraid I'll have to educate myself a little bit on #3. I hope that's alright.

Sure, feel free to ask. I am also not an expert on that topic, I just found the need while trying to read host certificates from a containerized app. More specific, I learned about p11-kit form [this flatpak discussion](https://github.com/p11-glue/p11-kit/issues/68), basically in flatpak they try to fetch hosts certificates but they face linux distro fragmentation issues, each distro cooks its own soup. Here is where p11-kit comes into play to provide an abstraction layer to access certificates. The implementation I did basically attempts to ask p11-kit and if it fails or returns an empty set it falls back to the current behavior of trying to find out certs at specific predefined locations.

File: src/index.js [hunk 1]
```diff
@@ -2,48 +2,82 @@ const es = require("event-stream");
...
+      try {
+        const urls = execSync(`p11tool ${ p11Args } ${ pkcsToken }`, {encoding: 'utf8'});
+        urls.split(EOL).forEach(line => {
+          if (!line){
+            return;
           }
+          const cert = execSync(`p11tool --export "${ line }"`, {encoding: 'utf8'});
+            certs.push(cert);
+        });
+        if (certs.lenght == 0){
```
Comment (kanishk98) [inline src/index.js]: Minor typo here, I think?
`if (certs.length === 0)`?

File: src/index.js [hunk 2]
```diff
@@ -2,48 +2,82 @@ const es = require("event-stream");
...
-          certs.push(file.split(splitPattern).map(cert => cert));
-          resolve(certs);
-        } catch (err) {
-          if (err.code !== "ENOENT") {
-            reject(err);
+      try {
+        const urls = execSync(`p11tool ${ p11Args } ${ pkcsToken }`, {encoding: 'utf8'});
+        urls.split(EOL).forEach(line => {
+          if (!line){
+            return;
```
Comment (kanishk98) [inline src/index.js:28]: Work for me: I'll add a git hook to prettify code automatically so no one has to worry about formatting. Didn't expect this tool to be used by anyone, tbh, good problems to have! :)

File: src/index.js [hunk 3]
```diff
@@ -2,48 +2,82 @@ const es = require("event-stream");
...
+            if (!line){
+              return;
             }
-            ++rejectedPaths;
-          } else {
-            certs.push(file.split(splitPattern).map(cert => cert));
-            resolve(certs);
+            certs.push(execSync(`p11tool --export "${ line }"`, {encoding: 'utf8'}));
+          }));
+          if (certs.lenght == 0){
```
Comment (kanishk98) [inline src/index.js]: Similar typo as above. :)

**Review by kanishk98 — CHANGES_REQUESTED** [review]

**Comment by kanishk98** [comment]
@davidcassany the PR looks mostly good to me, thank you so much for sending it in! I'm wondering if we can also add some test cases to this - feel free to nuke the current test files and replace them with yours if you have any ideas. Might be useful so we don't introduce regressions with later features?

**Comment by davidcassany** [comment]
> @davidcassany the PR looks mostly good to me, thank you so much for sending it in! I'm wondering if we can also add some test cases to this - feel free to nuke the current test files and replace them with yours if you have any ideas. Might be useful so we don't introduce regressions with later features?

I'll do my best to code some test for it. I hope it won't too hard, I have no experience on mocha. Meanwhile I realized that the `test.crt` file is missing in the repository, to me it feels this is not on purpose. I can try to add a `test.crt` of my choice,  but I am not sure if you had any special file for that. If you do don't I will just add some common cert for testing.

**Comment by davidcassany** [comment]
@kanishk98 Ok I finally added some tests based on the ones that were already in place. Basically I stubbed `child_process.exec` and `child_process.execSync` to mimic `p11tool` behavior and/or enforce looking for certificates on certain paths as fallback (as it was already).

**Comment by kanishk98** [comment]
@davidcassany thanks a ton for sending this in, and of course for your patience with the same! I realise you've been waiting for this release for a while, so let me reacquaint myself with publishing updates to an npm package and I'll push it out.
I clearly need to clean up the repo a little and make it easier to work with, will do that in a bit. Thank you!

**Comment by davidcassany** [comment]
You are absolutely welcome and thanks for taking care of it and merging -> releasing

**Pull request merged**